remotes/promises.go

```diff
@@ -0,0 +1,185 @@
+package remotes
+
+import (
+	"context"
+
+	"golang.org/x/sync/errgroup"
+)
+
+type scheduler interface {
+	schedule(func(ctx context.Context) error) promise
+	ctx() context.Context
+}
+
+type completionSource struct {
+	done chan struct{}
+	err  error
+}
+
+func (cs *completionSource) Done() <-chan struct{} {
+	return cs.done
+}
+
+func (cs *completionSource) Err() error {
+	return cs.err
+}
+
+func newPromise(scheduler scheduler, dependency dependency) promise {
+	return promise{scheduler: scheduler, dependency: dependency}
+}
+
+type promise struct {
+	dependency
+	scheduler scheduler
+}
+
+func (p promise) wait() error {
+	<-p.Done()
+	return p.Err()
+}
+
+func (p promise) then(next func(ctx context.Context) error) promise {
+	completionSource := &completionSource{
+		done: make(chan struct{}),
+	}
+	go func() {
+		defer close(completionSource.done)
+		select {
+		case <-p.Done():
+			if err := p.Err(); err != nil {
+				completionSource.err = err
+				return
+			}
+			completionSource.err = p.scheduler.schedule(next).wait()
+		}
+	}()
+	return newPromise(p.scheduler, completionSource)
+}
+
+type dependency interface {
+	Done() <-chan struct{}
+	Err() error
+}
+
+type failedDependency struct {
+	err error
+}
+
+func (f failedDependency) Done() <-chan struct{} {
+	return doneCh
+}
+
+func (f failedDependency) Err() error {
+	return f.err
+}
+
+type doneDependency struct {
+}
+
+func (doneDependency) Done() <-chan struct{} {
+	return doneCh
+}
+
+func (doneDependency) Err() error {
+	return nil
+}
+
+func whenAll(dependencies []dependency) dependency {
+	completionSource := &completionSource{
+		done: make(chan struct{}),
+	}
+	go func() {
+		defer close(completionSource.done)
+		gp, ctx := errgroup.WithContext(context.Background())
+		for _, dependency := range dependencies {
+			localDep := dependency
+			gp.Go(func() error {
```

Comment:
TODO: replace this 1 goroutine per dependency to a loop with reflect.Select (supporting arbitrary count of select cases).